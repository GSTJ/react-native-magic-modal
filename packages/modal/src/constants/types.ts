import Animated from "react-native-reanimated";

export type ModalChildren = React.FC;

export type Direction = "up" | "down" | "left" | "right";

export type ModalProps = {
  /**
   * Duration of the animation when the modal is shown.
   * @default 250
   */
  animationInTiming: number;

  /**
   * Duration of the animation when the modal is hidden.
   * @default 250
   */
  animationOutTiming: number;

  /**
   * If true, the backdrop will be hidden.
   * @default false
   */
  hideBackdrop: boolean;

  /**
   * The color of the backdrop.
   * @default "rgba(0, 0, 0, 0.5)"
   */
  backdropColor: string;

  /**
   * Function to be called when the back button is pressed.
   * You can override it to prevent the modal from closing on back button press.
   * @default undefined
   * @example () => { console.log('Back button pressed'); magicModal.hide(); }
   */
  onBackButtonPress: (() => void) | undefined;

  /**
   * Function to be called when the backdrop is pressed.
   * You can override it to prevent the modal from closing on backdrop press.
   * @default undefined
   * @example () => { console.log('Backdrop pressed'); magicModal.hide(); }
   */
  onBackdropPress: (() => void) | undefined;

  /**
   * Custom style for the modal.
   * @default {}
   * @example { backgroundColor: 'red', padding: 10 }
   */
  style: Record<string, unknown>;

  /**
   * Damping factor for the swipe gesture.
   * @default 0.2
   */
  dampingFactor: number;

  /**
   * Direction of the modal animation.
   * Set to undefined to disable the swipe gesture.
   * @default "down"
   * @example "up"
   */
  swipeDirection: Direction | undefined;

  /**
   * Velocity threshold for the swipe gesture.
   * @default 500
   */
  swipeVelocityThreshold: number;
} & Pick<React.ComponentProps<typeof Animated.View>, "entering" | "exiting">;

// eslint-disable-next-line @typescript-eslint/no-explicit-any
export type GenericFunction = (props: any) => any;

export type GlobalHideFunction = (
  props?: unknown,
  options?: { modalID?: string },
) => void;

export type HookHideFunction = (props?: unknown) => void;

export type NewConfigProps = Partial<ModalProps>;

export type GlobalShowFunction = <T>(
  newComponent: ModalChildren,
  newConfig?: NewConfigProps,
) => { promise: Promise<T | undefined>; modalID: string };

export enum MagicModalHideTypes {
  BACKDROP_PRESSED = "BACKDROP_PRESSED",
  SWIPE_COMPLETED = "SWIPE_COMPLETED",
  BACK_BUTTON_PRESSED = "BACK_BUTTON_PRESSED",
  MODAL_OVERRIDE = "MODAL_OVERRIDE",
}
