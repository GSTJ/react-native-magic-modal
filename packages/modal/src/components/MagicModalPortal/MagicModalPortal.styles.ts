import { StyleSheet } from "react-native";

export const styles = StyleSheet.create({
  container: {
    margin: 0,
  },
  overlay: {
    flex: 1,
    justifyContent: "center",
  },
  backdrop: {
    flex: 1,
    backgroundColor: "rgba(0, 0, 0, 0.5)",
  },
  backdropContainer: {
    position: "absolute",
    width: "100%",
    height: "100%",
  },
});
