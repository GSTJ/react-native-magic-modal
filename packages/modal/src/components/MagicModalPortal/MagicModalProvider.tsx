import React, { memo, useMemo } from "react";
import { HookHideFunction } from "../../constants/types";

const MagicModalContext = React.createContext<{
  hide: HookHideFunction;
}>({
  hide: async () => {},
});

export const useMagicModal = () => {
  const context = React.useContext(MagicModalContext);

  if (!context) {
    throw new Error("useMagicModal must be used within a MagicModalProvider");
  }

  return context;
};

export const MagicModalProvider = memo(
  ({
    children,
    hide,
  }: {
    children: React.ReactNode;
    hide: HookHideFunction;
  }) => {
    const value = useMemo(() => ({ hide }), [hide]);

    return (
      <MagicModalContext.Provider value={value}>
        {children}
      </MagicModalContext.Provider>
    );
  },
);
