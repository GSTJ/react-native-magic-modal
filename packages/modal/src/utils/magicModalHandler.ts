import React from "react";

import {
  GlobalHideAllFunction,
  GlobalHideFunction,
  GlobalShowFunction,
  // HideReturn is used in JS Doc
  // eslint-disable-next-line unused-imports/no-unused-imports, @typescript-eslint/no-unused-vars
  HideReturn,
} from "../constants/types";

export const magicModalRef = React.createRef<IModal>();

const getMagicModal = (): NonNullable<typeof magicModalRef.current> => {
  if (!magicModalRef.current) {
    throw new Error(
      "MagicModalPortal not found. Please wrap your component with MagicModalPortal.",
    );
  }
  return magicModalRef.current;
};

const show: GlobalShowFunction = (newComponent, newConfig) => {
  return getMagicModal().show(newComponent, newConfig);
};

const hide: GlobalHideFunction = (props, { modalID } = {}) => {
  return getMagicModal().hide(props, { modalID });
};

const hideAll: GlobalHideAllFunction = () => {
  // We recommend using this method in jest, and having throw because the ref was not found isn't useful there.
  // Not all tests are necessarily using the provider.
  return magicModalRef.current?.hideAll();
};
export interface IModal {
  show: typeof show;
  hide: typeof hide;
  hideAll: typeof hideAll;
}

/**
 * @example
 * ```js
 * // ...
 * import { magicModal } from 'react-native-magic-toast';
 *
 * // ...
 * const ExampleModal = () => (
 *  const { hide } = useMagicModal<{ message: string }>();
 *  <TouchableOpacity onPress={() => hide({ message: "hey" })}>
 *    <Text>Test!</Text>
 *  </TouchableOpacity>
 * )
 *
 * const result = magicModal.show(ExampleModal);
 * console.log(await result.promise); // Returns { reason: MagicModalHideReason.INTENTIONAL_HIDE, message: "hey" } when the modal is closed by the TouchableOpacity.
 * ```
 */
export const magicModal = {
  /**
   * @description Pushes a modal to the Stack, it will be displayed on top of the others.
   * @param newComponent Recieves a function that returns a modal component.
   * @param newConfig Recieves {@link NewConfigProps}  to override the default configs.
   * @returns Returns a Promise that resolves with the {@link hide} props when the Modal is closed. If it were closed automatically, without the manual use of {@link hide}, the return would be one of {@link HideReturn}
   */
  show,
  /**
   * @description Hides the given modal. Prefer using `hide` from `useMagicModal`, as it already infers the modalID.
   * You should use the `magicModal.hide` function directly  only when calling from outside the modal.
   * @param props Those props will be passed to the {@link show} resolve function.
   */
  hide,
  /**
   * @description Hides all modals in the stack. This function should be used sparingly, as it's generally preferable to hide modals individually from within the modal itself.
   * However, this function can be useful in edge cases. It's also useful for test suites, such as calling hideAll in Jest's beforeEach function as a cleanup step.
   */
  hideAll,
};
