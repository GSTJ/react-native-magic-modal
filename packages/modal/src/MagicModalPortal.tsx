import React, {
  useCallback,
  useImperativeHandle,
  useMemo,
  useRef,
  useState,
} from "react";
import Animated, {
  Extrapolation,
  interpolate,
  runOnJS,
  useAnimatedStyle,
  useSharedValue,
  withSpring,
} from "react-native-reanimated";
import {
  Gesture,
  GestureDetector,
  GestureHandlerRootView,
} from "react-native-gesture-handler";

import { ANIMATION_DURATION_IN_MS } from "./constants/animations";
import type { IModal, ModalChildren } from "./utils/magicModalHandler";
import { magicModalRef } from "./utils/magicModalHandler";
import { styles } from "./MagicModalPortal.styles";
import { FullWindowOverlay } from "react-native-screens";
import { Pressable, StyleSheet, useWindowDimensions } from "react-native";

export type Direction = "top" | "bottom" | "left" | "right";

export type ModalProps = {
  /**
   * Duration of the animation when the modal is shown.
   * @default 300
   */
  animationInTiming: number;

  /**
   * Duration of the animation when the modal is hidden.
   * @default 300
   */
  animationOutTiming: number;

  /**
   * If true, the backdrop will be hidden.
   * @default false
   */
  hideBackdrop: boolean;

  /**
   * The color of the backdrop.
   * @default "rgba(0, 0, 0, 0.5)"
   */
  backdropColor: string;

  /**
   * Function to be called when the back button is pressed.
   * @default undefined
   * @example () => { console.log('Back button pressed'); magicModal.hide(); }
   */
  onBackButtonPress: (() => void) | undefined;

  /**
   * Function to be called when the backdrop is pressed.
   * @default undefined
   * @example () => { console.log('Backdrop pressed'); magicModal.hide(); }
   */
  onBackdropPress: (() => void) | undefined;

  /**
   * Custom style for the modal.
   * @default {}
   * @example { backgroundColor: 'red', padding: 10 }
   */
  style: Record<string, unknown>;

  /**
   * Damping factor for the swipe gesture.
   * @default 0.2
   */
  dampingFactor: number;

  /**
   * Direction of the modal animation.
   * @default "bottom"
   * @example "top"
   */
  direction: Direction;

  /**
   * Velocity threshold for the swipe gesture.
   * @default 500
   */
  swipeVelocityThreshold: number;
};

type GenericFunction = (props: any) => any;

export enum MagicModalHideTypes {
  BACKDROP_PRESSED = "BACKDROP_PRESSED",
  SWIPE_COMPLETED = "SWIPE_COMPLETED",
  BACK_BUTTON_PRESSED = "BACK_BUTTON_PRESSED",
  MODAL_OVERRIDE = "MODAL_OVERRIDE",
}

export const modalRefForTests = React.createRef<any>();

const AnimatedPressable = Animated.createAnimatedComponent(Pressable);

const defaultConfig = {
  animationInTiming: ANIMATION_DURATION_IN_MS,
  animationOutTiming: ANIMATION_DURATION_IN_MS,
  hideBackdrop: false,
  backdropColor: "rgba(0, 0, 0, 0.5)",
  dampingFactor: 0.2,
  direction: "bottom",
  swipeVelocityThreshold: 500,
  onBackButtonPress: undefined,
  onBackdropPress: undefined,
  style: {},
} satisfies ModalProps;

/**
 * @description A magic portal that should stay on the top of the app component hierarchy for the modal to be displayed.
 * @example
 * ```js
 * import { MagicModalPortal } from 'react-native-magic-modal';
 *
 * export default function App() {
 *   return (
 *     <SomeRandomProvider>
 *       <MagicModalPortal />  // <-- On the top of the app component hierarchy
 *       <Router /> // Your app router or something could follow below
 *     </SomeRandomProvider>
 *   );
 * }
 * ```
 */
export const MagicModalPortal: React.FC = () => {
  const [isVisible, setIsVisible] = useState(false);
  const [config, setConfig] = useState<ModalProps>(defaultConfig);
  const [modalContent, setModalContent] = useState<React.ReactNode>(<></>);

  const translationX = useSharedValue(0);
  const translationY = useSharedValue(0);
  const prevTranslationX = useSharedValue(0);
  const prevTranslationY = useSharedValue(0);

  const onHideRef = useRef<GenericFunction>(() => {});

  const { width, height } = useWindowDimensions();

  const getDirectionTranslation = (direction: Direction, value: number) => {
    switch (direction) {
      case "left":
        return { translationX: -value, translationY: 0 };
      case "right":
        return { translationX: value, translationY: 0 };
      case "top":
        return { translationX: 0, translationY: -value };
      case "bottom":
        return { translationX: 0, translationY: value };
    }
  };

  const hide = useCallback<IModal["hide"]>(
    async (props) => {
      const springConfig = {
        duration: config.animationOutTiming,
        dampingRatio: 3,
      };

      const directionTranslation = getDirectionTranslation(
        config.direction,
        height
      );

      await new Promise<void>((resolve) => {
        if (directionTranslation.translationX !== 0) {
          translationX.value = withSpring(
            directionTranslation.translationX,
            springConfig,
            () => runOnJS(resolve)()
          );
          return;
        }

        translationY.value = withSpring(
          directionTranslation.translationY,
          springConfig,
          () => runOnJS(resolve)()
        );
      });

      setIsVisible(false);

      onHideRef.current(props);
    },
    [
      config.animationOutTiming,
      config.direction,
      height,
      translationX,
      translationY,
    ]
  );

  useImperativeHandle(magicModalRef, () => ({
    hide,
    show: async (
      newComponent: ModalChildren,
      newConfig: Partial<ModalProps> = {}
    ) => {
      if (isVisible) await hide(MagicModalHideTypes.MODAL_OVERRIDE);

      const springConfig = {
        duration: config.animationInTiming,
        dampingRatio: 1,
      };

      const startPosition = {
        bottom: {
          translationX: 0,
          translationY: height,
        },
        top: {
          translationX: 0,
          translationY: -height,
        },
        left: {
          translationX: -width,
          translationY: 0,
        },
        right: {
          translationX: width,
          translationY: 0,
        },
      };

      const mergedConfig = {
        ...defaultConfig,
        ...newConfig,
      };

      translationX.value = startPosition[mergedConfig.direction].translationX;
      translationY.value = startPosition[mergedConfig.direction].translationY;

      translationX.value = withSpring(0, springConfig);
      translationY.value = withSpring(0, springConfig);

      prevTranslationX.value = 0;
      prevTranslationY.value = 0;

      setModalContent(newComponent as unknown as React.ReactNode);
      setConfig(mergedConfig);
      setIsVisible(true);

      return new Promise((resolve) => {
        onHideRef.current = resolve;
      });
    },
  }));

  const onBackdropPress = useMemo(() => {
    return (
      config.onBackdropPress ??
      (() => hide(MagicModalHideTypes.BACKDROP_PRESSED))
    );
  }, [config.onBackdropPress, hide]);

  const pan = Gesture.Pan()
    .minDistance(1)
    .onStart(() => {
      prevTranslationX.value = translationX.value;
      prevTranslationY.value = translationY.value;
    })
    .onUpdate((event) => {
      const direction = config.direction;
      const translationValue =
        direction.includes("left") || direction.includes("right")
          ? event.translationX
          : event.translationY;

      const prevTranslationValue =
        direction.includes("left") || direction.includes("right")
          ? prevTranslationX.value
          : prevTranslationY.value;

      const shouldDamp =
        (direction === "bottom" && translationValue < 0) ||
        (direction === "top" && translationValue > 0) ||
        (direction === "left" && translationValue > 0) ||
        (direction === "right" && translationValue < 0);

      const dampedTranslation = shouldDamp
        ? prevTranslationValue + translationValue * config.dampingFactor
        : prevTranslationValue + translationValue;

      const maxTranslate =
        direction.includes("left") || direction.includes("right")
          ? width / 2 - 50
          : height / 2 - 50;

      const limitedTranslation =
        Math.abs(dampedTranslation) > Math.abs(maxTranslate)
          ? maxTranslate * Math.sign(dampedTranslation)
          : dampedTranslation;

      if (direction.includes("left") || direction.includes("right")) {
        translationX.value = limitedTranslation;
      } else {
        translationY.value = limitedTranslation;
      }
    })
    .onEnd((event) => {
      const velocityThreshold = config.swipeVelocityThreshold;

      const shouldHide =
        (config.direction === "right" && event.velocityX > velocityThreshold) ||
        (config.direction === "left" && event.velocityX < -velocityThreshold) ||
        (config.direction === "top" && event.velocityY < -velocityThreshold) ||
        (config.direction === "bottom" && event.velocityY > velocityThreshold);

      if (shouldHide) {
        hide(MagicModalHideTypes.SWIPE_COMPLETED);
        return;
      }

      translationX.value = withSpring(0, {
        velocity: event.velocityX,
        damping: 75,
      });
      translationY.value = withSpring(0, {
        velocity: event.velocityY,
        damping: 75,
      });
    })
    .runOnJS(true);

  const animatedStyles = useAnimatedStyle(() => ({
    transform: [
      { translateX: translationX.value },
      { translateY: translationY.value },
    ],
  }));

  const animatedBackdropStyles = useAnimatedStyle(() => {
    const translationValue =
      config.direction.includes("left") || config.direction.includes("right")
        ? translationX.value
        : translationY.value;

    const rangeMap = {
      left: [-width, 0],
      right: [width, 0],
      top: [-height, 0],
      bottom: [height, 0],
    };

    return {
      opacity: interpolate(
        translationValue,
        rangeMap[config.direction],
        [0, 1],
        Extrapolation.CLAMP
      ),
    };
  });

  return (
    <FullWindowOverlay>
      <GestureHandlerRootView
        pointerEvents="box-none"
        style={StyleSheet.absoluteFill}
      >
        {isVisible && (
          <>
            <AnimatedPressable
              pointerEvents={config.hideBackdrop ? "none" : "auto"}
              style={[
                styles.backdrop,
                animatedBackdropStyles,
                {
                  backgroundColor: config.hideBackdrop
                    ? "transparent"
                    : config.backdropColor ?? styles.backdrop.backgroundColor,
                },
              ]}
              onPress={onBackdropPress}
            />
            <Animated.View
              pointerEvents="box-none"
              style={[
                animatedStyles,
                styles.overlay,
                styles.container,
                config.style,
              ]}
            >
              <GestureDetector gesture={pan}>{modalContent}</GestureDetector>
            </Animated.View>
          </>
        )}
      </GestureHandlerRootView>
    </FullWindowOverlay>
  );
};
