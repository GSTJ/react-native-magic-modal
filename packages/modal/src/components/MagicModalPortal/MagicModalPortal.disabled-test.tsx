// TODO: Re-enable tests when we figure out the jest issues on the project.
import React from "react";
import { Text } from "react-native";
import { render, waitFor } from "@testing-library/react-native";

import { magicModal } from "../../utils/magicModalHandler";
import {
  MagicModalHideTypes,
  MagicModalPortal,
  modalRefForTests,
} from "./MagicModalPortal";
import { setUpTests } from "react-native-reanimated";

jest.mock("react-native-screens", () => ({
  FullWindowOverlay: ({ children }: { children: React.ReactNode }) => (
    <>{children}</>
  ),
}));

setUpTests();

describe("MagicModal", () => {
  it("renders correctly", async () => {
    const component = render(<MagicModalPortal />);

    expect(component).toMatchSnapshot();
  });

  it("renders children conditionally", () => {
    const component = render(<MagicModalPortal />);

    const testId = "children-magic-modal";

    expect(component.queryByTestId(testId)).toBeFalsy();

    magicModal.show(() => <Text testID={testId}>Taveira</Text>);

    expect(component.queryByTestId(testId)).toBeTruthy();
    expect(component).toMatchSnapshot();
  });

  it("should redirect hide params as show promise result", async () => {
    render(<MagicModalPortal />);

    const modalResultPromise = magicModal.show(() => <Text>Taveira</Text>);

    magicModal.hide("some-result-2");

    const modalResult = await modalResultPromise;
    expect(modalResult).toBe("some-result-2");
  });

  it("should override old modal on show", async () => {
    const component = render(<MagicModalPortal />);

    const oldModalRetunPromise = magicModal.show(() => (
      <Text testID="old-modal">Taveira</Text>
    ));

    await waitFor(() => {
      expect(component.queryByTestId("old-modal")).toBeTruthy();
      expect(component.queryByTestId("new-modal")).toBeFalsy();
    });

    magicModal.show(() => <Text testID="new-modal">Taveira</Text>);

    expect(await oldModalRetunPromise).toBe(MagicModalHideTypes.MODAL_OVERRIDE);

    await waitFor(() => {
      expect(component.queryByTestId("new-modal")).toBeTruthy();
      expect(component.queryByTestId("old-modal")).toBeFalsy();
    });
  });

  describe("should return one of MagicModalHideTypes on automatic hides", () => {
    let modalResultPromise: any = null;

    beforeEach(async () => {
      render(<MagicModalPortal />);

      modalResultPromise = magicModal.show(() => (
        <Text testID="my-modal">Taveira</Text>
      ));
    });

    it("should return `MagicModalHideTypes.BACK_BUTTON_PRESSED` when the back button is pressed", async () => {
      modalRefForTests.current.props.onBackButtonPress();

      expect(await modalResultPromise).toBe(
        MagicModalHideTypes.BACK_BUTTON_PRESSED
      );
    });

    it("should return `MagicModalHideTypes.BACKDROP_PRESSED` when the backdrop is pressed", async () => {
      modalRefForTests.current.props.onBackdropPress();

      expect(await modalResultPromise).toBe(
        MagicModalHideTypes.BACKDROP_PRESSED
      );
    });

    it("should return `MagicModalHideTypes.SWIPE_COMPLETED` when the swipe is complete", async () => {
      modalRefForTests.current.props.onSwipeComplete();

      expect(await modalResultPromise).toBe(
        MagicModalHideTypes.SWIPE_COMPLETED
      );
    });
  });
});
