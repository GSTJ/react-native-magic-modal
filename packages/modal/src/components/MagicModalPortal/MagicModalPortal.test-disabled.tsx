import React from "react";
import { Text } from "react-native";
import { render } from "@testing-library/react-native";

import { MagicModalHideReason } from "../../constants/types";
import { magicModal } from "../../utils/magicModalHandler";
import { MagicModalPortal } from "./MagicModalPortal";

describe("MagicModal", () => {
  it("renders correctly", async () => {
    const component = render(<MagicModalPortal />);

    expect(component).toMatchSnapshot();
  });

  it("renders children conditionally", () => {
    const component = render(<MagicModalPortal />);

    const testId = "children-magic-modal";

    expect(component.queryByTestId(testId)).toBeFalsy();

    magicModal.show(() => <Text testID={testId}>Taveira</Text>);

    expect(component.queryByTestId(testId)).toBeTruthy();
    expect(component).toMatchSnapshot();
  });

  it("should redirect hide params as show promise result", async () => {
    render(<MagicModalPortal />);

    const modalResultPromise = magicModal.show(() => <Text>Taveira</Text>);

    magicModal.hide("some-result-2");

    const modalResult = await modalResultPromise;
    expect(modalResult).toBe("some-result-2");
  });

  describe("should return one of MagicModalHideTypes on automatic hides", () => {
    let modalResultPromise: unknown = null;

    beforeEach(async () => {
      render(<MagicModalPortal />);

      modalResultPromise = magicModal.show(() => (
        <Text testID="my-modal">Taveira</Text>
      ));
    });

    it("should return `MagicModalHideTypes.BACK_BUTTON_PRESS` when the back button is press", async () => {
      // modalRefForTests.current.props.onBackButtonPress();

      expect(await modalResultPromise).toBe({
        reason: MagicModalHideReason.BACK_BUTTON_PRESS,
      });
    });

    it("should return `MagicModalHideTypes.BACKDROP_PRESS` when the backdrop is press", async () => {
      // modalRefForTests.current.props.onBackdropPress();

      expect(await modalResultPromise).toBe({
        reason: MagicModalHideReason.BACKDROP_PRESS,
      });
    });

    it("should return `MagicModalHideTypes.SWIPE_COMPLETE` when the swipe is complete", async () => {
      // modalRefForTests.current.props.onSwipeComplete();

      expect(await modalResultPromise).toBe({
        reason: MagicModalHideReason.SWIPE_COMPLETE,
      });
    });
  });
});
