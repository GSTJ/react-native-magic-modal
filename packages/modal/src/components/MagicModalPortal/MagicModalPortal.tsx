import React, {
  memo,
  useCallback,
  useEffect,
  useImperativeHandle,
  useMemo,
  useRef,
  useState,
} from "react";
import Animated, {
  Extrapolation,
  FadeIn,
  FadeInDown,
  FadeInLeft,
  FadeInRight,
  FadeInUp,
  FadeOut,
  FadeOutDown,
  FadeOutLeft,
  FadeOutRight,
  FadeOutUp,
  interpolate,
  runOnJS,
  useAnimatedStyle,
  useSharedValue,
  withSpring,
} from "react-native-reanimated";
import { Gesture, GestureDetector } from "react-native-gesture-handler";
import type { IModal } from "../../utils/magicModalHandler";
import { magicModalRef } from "../../utils/magicModalHandler";
import { styles } from "./MagicModalPortal.styles";
import { FullWindowOverlay } from "../FullWindowOverlay/FullWindowOverlay";
import {
  BackHandler,
  Pressable,
  StyleSheet,
  View,
  useWindowDimensions,
} from "react-native";
import {
  ModalProps,
  GenericFunction,
  MagicModalHideTypes,
  ModalChildren,
  Direction,
} from "../../constants/types";
import { defaultConfig, defaultDirection } from "../../constants/defaultConfig";

export const modalRefForTests = React.createRef<any>();

const AnimatedPressable = Animated.createAnimatedComponent(Pressable);

const defaultAnimationInMap = {
  up: FadeInUp,
  down: FadeInDown,
  left: FadeInLeft,
  right: FadeInRight,
} satisfies Record<Direction, unknown>;

const defaultAnimationOutMap = {
  up: FadeOutUp,
  down: FadeOutDown,
  left: FadeOutLeft,
  right: FadeOutRight,
} satisfies Record<Direction, unknown>;

/**
 * @description A magic portal that should stay on the top of the app component hierarchy for the modal to be displayed.
 * @example
 * ```js
 * import { MagicModalPortal } from 'react-native-magic-modal';
 *
 * export default function App() {
 *   return (
 *     <SomeRandomProvider>
 *       <MagicModalPortal />  // <-- On the top of the app component hierarchy
 *       <Router /> // Your app router or something could follow below
 *     </SomeRandomProvider>
 *   );
 * }
 * ```
 */
export const MagicModalPortal: React.FC = memo(() => {
  const [config, setConfig] = useState<ModalProps>(defaultConfig);
  const [modalContent, setModalContent] = useState<React.ReactNode>(undefined);

  const translationX = useSharedValue(0);
  const translationY = useSharedValue(0);
  const prevTranslationX = useSharedValue(0);
  const prevTranslationY = useSharedValue(0);

  const onHideRef = useRef<GenericFunction>(() => {});

  const { width, height } = useWindowDimensions();

  const hide = useCallback<IModal["hide"]>(async (props) => {
    setModalContent(undefined);

    await new Promise((resolve) => {
      setTimeout(resolve, config.animationOutTiming);
    });

    translationX.value = 0;
    translationY.value = 0;

    onHideRef.current(props);
  }, []);

  useImperativeHandle(magicModalRef, () => ({
    hide,
    show: async (
      newComponent: ModalChildren,
      newConfig: Partial<ModalProps> = {},
    ) => {
      if (modalContent) await hide(MagicModalHideTypes.MODAL_OVERRIDE);

      const mergedConfig = { ...defaultConfig, ...newConfig };

      setConfig(mergedConfig);
      setModalContent(newComponent as unknown as React.ReactNode);

      return new Promise((resolve) => {
        onHideRef.current = resolve;
      });
    },
  }));

  const onBackdropPress = useMemo(() => {
    return (
      config.onBackdropPress ??
      (() => hide(MagicModalHideTypes.BACKDROP_PRESSED))
    );
  }, [config.onBackdropPress, hide]);

  const isHorizontal =
    config.swipeDirection === "left" || config.swipeDirection === "right";

  const rangeMap = useMemo(
    () =>
      ({
        up: -height,
        down: height,
        left: -width,
        right: width,
      }) satisfies Record<Direction, number>,
    [height, width],
  );

  const pan = Gesture.Pan()
    .enabled(!!config.swipeDirection)
    .minDistance(1)
    .onStart(() => {
      prevTranslationX.value = translationX.value;
      prevTranslationY.value = translationY.value;
    })
    .onUpdate((event) => {
      const translationValue = isHorizontal
        ? event.translationX
        : event.translationY;

      const prevTranslationValue = isHorizontal
        ? prevTranslationX.value
        : prevTranslationY.value;

      const shouldDampMap = {
        up: translationValue > 0,
        down: translationValue < 0,
        left: translationValue > 0,
        right: translationValue < 0,
      } satisfies Record<Direction, boolean>;

      const shouldDamp =
        shouldDampMap[config.swipeDirection ?? defaultDirection];

      const dampedTranslation = shouldDamp
        ? prevTranslationValue + translationValue * config.dampingFactor
        : prevTranslationValue + translationValue;

      if (isHorizontal) {
        translationX.value = dampedTranslation;
      } else {
        translationY.value = dampedTranslation;
      }
    })
    .onEnd((event) => {
      const velocityThreshold = config.swipeVelocityThreshold;

      const shouldHideMap = {
        up: event.velocityY < -velocityThreshold,
        down: event.velocityY > velocityThreshold,
        right: event.velocityX > velocityThreshold,
        left: event.velocityX < -velocityThreshold,
      } satisfies Record<Direction, boolean>;

      const shouldHide =
        shouldHideMap[config.swipeDirection ?? defaultDirection];

      if (!shouldHide) {
        translationX.value = withSpring(0, {
          velocity: event.velocityX,
          damping: 75,
        });
        translationY.value = withSpring(0, {
          velocity: event.velocityY,
          damping: 75,
        });
        return;
      }

      const mainTranslation = isHorizontal ? translationX : translationY;

      mainTranslation.value = withSpring(
        rangeMap[config.swipeDirection ?? defaultDirection],
        { velocity: event.velocityX, overshootClamping: true },
        (success) =>
          success && runOnJS(hide)(MagicModalHideTypes.SWIPE_COMPLETED),
      );

      return;
    });

  const animatedStyles = useAnimatedStyle(() => ({
    transform: [
      { translateX: translationX.value },
      { translateY: translationY.value },
    ],
  }));

  const animatedBackdropStyles = useAnimatedStyle(() => {
    const translationValue = isHorizontal
      ? translationX.value
      : translationY.value;

    return {
      opacity: interpolate(
        translationValue,
        [rangeMap[config.swipeDirection ?? defaultDirection], 0],
        [0, 1],
        Extrapolation.CLAMP,
      ),
    };
  });

  useEffect(() => {
    if (!modalContent) return;
    const backHandler = BackHandler.addEventListener(
      "hardwareBackPress",
      () => {
        if (config.onBackButtonPress) {
          config.onBackButtonPress();
        } else {
          hide(MagicModalHideTypes.BACK_BUTTON_PRESSED);
        }
        return true;
      },
    );
    return () => backHandler.remove();
  }, [config.onBackButtonPress, hide, modalContent]);

  const isBackdropVisible = modalContent && !config.hideBackdrop;

  return (
    <FullWindowOverlay>
      {/* This needs to always be rendered, if we make it conditionally render based on ModalContent too,
          the modal will have zIndex issues on react-navigation modals. */}
      <View pointerEvents="box-none" style={StyleSheet.absoluteFill}>
        {modalContent ? (
          <>
            <Animated.View
              pointerEvents={isBackdropVisible ? "auto" : "none"}
              entering={FadeIn.duration(config.animationInTiming)}
              exiting={FadeOut.duration(config.animationOutTiming)}
              style={styles.backdropContainer}
            >
              <AnimatedPressable
                style={[
                  styles.backdrop,
                  animatedBackdropStyles,
                  {
                    backgroundColor: isBackdropVisible
                      ? config.backdropColor
                      : "transparent",
                  },
                ]}
                onPress={onBackdropPress}
              />
            </Animated.View>
            <Animated.View
              pointerEvents="box-none"
              style={[styles.overlay, animatedStyles]}
            >
              <Animated.View
                pointerEvents="box-none"
                style={[styles.overlay, config.style]}
                entering={
                  config.entering ??
                  defaultAnimationInMap[
                    config.swipeDirection ?? defaultDirection
                  ].duration(config.animationInTiming)
                }
                exiting={
                  config.exiting ??
                  defaultAnimationOutMap[
                    config.swipeDirection ?? defaultDirection
                  ].duration(config.animationOutTiming)
                }
              >
                <GestureDetector gesture={pan}>{modalContent}</GestureDetector>
              </Animated.View>
            </Animated.View>
          </>
        ) : null}
      </View>
    </FullWindowOverlay>
  );
});
