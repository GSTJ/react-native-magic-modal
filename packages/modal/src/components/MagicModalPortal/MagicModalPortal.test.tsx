import React from "react";
import { Text } from "react-native";
import { act, render } from "@testing-library/react-native";

import { MagicModalHideReason } from "../../constants/types";
import { magicModal } from "../../utils/magicModalHandler";
import { MagicModalPortal } from "./MagicModalPortal";

describe("MagicModalPortal", () => {
  afterEach(() => {
    jest.clearAllMocks();
  });

  it("renders correctly", async () => {
    const component = render(<MagicModalPortal />);
    expect(component).toMatchSnapshot();
  });

  it("renders children conditionally", async () => {
    const component = render(<MagicModalPortal />);
    const testId = "children-magic-modal";

    expect(component.queryByTestId(testId)).toBeFalsy();

    await act(async () => {
      magicModal.show(() => <Text testID={testId}>Taveira</Text>);
    });

    expect(component.queryByTestId(testId)).toBeTruthy();
  });

  it("redirects hide params to promise result data", async () => {
    render(<MagicModalPortal />);

    let modalResult;
    await act(async () => {
      const { promise } = magicModal.show(() => <Text>Taveira</Text>);

      // Store the promise for later resolution
      modalResult = promise;

      // Hide with specific data
      magicModal.hide("some-result-2");
    });

    // The promise should resolve with the hide params in the data field
    expect(await modalResult).toEqual({
      reason: MagicModalHideReason.INTENTIONAL_HIDE,
      data: "some-result-2",
    });
  });

  describe("returns appropriate hide reasons", () => {
    it("returns INTENTIONAL_HIDE when explicitly hidden", async () => {
      render(<MagicModalPortal />);

      let modalPromise;
      await act(async () => {
        const { promise } = magicModal.show(() => (
          <Text testID="my-modal">Taveira</Text>
        ));

        modalPromise = promise;

        // Explicitly hide with data
        magicModal.hide("test-data");
      });

      const result = await modalPromise;
      expect(result).toEqual({
        reason: MagicModalHideReason.INTENTIONAL_HIDE,
        data: "test-data",
      });
    });
  });
});
