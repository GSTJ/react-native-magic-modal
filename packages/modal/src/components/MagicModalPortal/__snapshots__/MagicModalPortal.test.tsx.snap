// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MagicModalPortal renders correctly 1`] = `
<RNSFullWindowOverlay
  style={
    [
      {
        "bottom": 0,
        "left": 0,
        "position": "absolute",
        "right": 0,
        "top": 0,
      },
      {
        "height": 1334,
        "width": 750,
      },
    ]
  }
>
  <View
    style={
      [
        {
          "bottom": 0,
          "left": 0,
          "position": "absolute",
          "right": 0,
          "top": 0,
        },
        {
          "pointerEvents": "box-none",
        },
      ]
    }
  />
</RNSFullWindowOverlay>
`;
